src/kokkos/plugin-SiPixelClusterizer/kokkos/gpuClusterChargeCut.h

```diff
@@ -4,126 +4,149 @@
+                id(i) = ::gpuClustering::InvId;
+                clusterId(i) = ::gpuClustering::InvId;
+              }
+            }
+            nclus = ::gpuClustering::MaxNumClustersPerModules;
           }
-        }
-        nclus = MaxNumClustersPerModules;
-      }
 
 #ifdef GPU_DEBUG
-      if (thisModuleId % 100 == 1)
-        if (threadIdx.x == 0)
-          printf("start clusterizer for module %d in block %d\n", thisModuleId, blockIdx.x);
+          if (thisModuleId % 100 == 1)
+            if (teamMember.team_rank() == 0)
+              printf("start clusterizer for module %d in block %d\n", thisModuleId, teamMember.league_rank());
 #endif
 
-      __shared__ int32_t charge[MaxNumClustersPerModules];
-      __shared__ uint8_t ok[MaxNumClustersPerModules];
-      __shared__ uint16_t newclusId[MaxNumClustersPerModules];
-
-      assert(nclus <= MaxNumClustersPerModules);
-      for (auto i = threadIdx.x; i < nclus; i += blockDim.x) {
-        charge[i] = 0;
-      }
-      __syncthreads();
-
-      for (auto i = first; i < numElements; i += blockDim.x) {
-        if (id[i] == InvId)
-          continue;  // not valid
-        if (id[i] != thisModuleId)
-          break;  // end of module
-        atomicAdd(&charge[clusterId[i]], adc[i]);
-      }
-      __syncthreads();
-
-      auto chargeCut = thisModuleId < 96 ? 2000 : 4000;  // move in constants (calib?)
-      for (auto i = threadIdx.x; i < nclus; i += blockDim.x) {
-        newclusId[i] = ok[i] = charge[i] > chargeCut ? 1 : 0;
-      }
-
-      __syncthreads();
-
-      // renumber
-      __shared__ uint16_t ws[32];
-      blockPrefixScan(newclusId, nclus, ws);
-
-      assert(nclus >= newclusId[nclus - 1]);
-
-      if (nclus == newclusId[nclus - 1])
-        return;
-
-      nClustersInModule[thisModuleId] = newclusId[nclus - 1];
-      __syncthreads();
-
-      // mark bad cluster again
-      for (auto i = threadIdx.x; i < nclus; i += blockDim.x) {
-        if (0 == ok[i])
-          newclusId[i] = InvId + 1;
-      }
-      __syncthreads();
-
-      // reassign id
-      for (auto i = first; i < numElements; i += blockDim.x) {
-        if (id[i] == InvId)
-          continue;  // not valid
-        if (id[i] != thisModuleId)
-          break;  // end of module
-        clusterId[i] = newclusId[clusterId[i]] - 1;
-        if (clusterId[i] == InvId)
-          id[i] = InvId;
-      }
+          charge_view_type charge(teamMember.team_scratch(shared_view_level),::gpuClustering::MaxNumClustersPerModules);
+          ok_view_type ok(teamMember.team_scratch(shared_view_level),::gpuClustering::MaxNumClustersPerModules);
+          newclusid_view_type newclusId(teamMember.team_scratch(shared_view_level),::gpuClustering::MaxNumClustersPerModules);
+
+          assert(nclus <= ::gpuClustering::MaxNumClustersPerModules);
+          for (uint32_t i = teamMember.team_rank(); i < ::gpuClustering::MaxNumClustersPerModules; i += teamMember.team_size()) {
```

Comment:
Since this loop is initializing the charge array to 0. I decided to loop over the full array (and initialize all the arrays) instead of using `nclus`. I think `nclus` is typically set to the `MaxNumClustersPerModules` but I thought it safer to just initialize everything across the full memory space.

Reply:
Ok